projects/extension/src/background/ConnectionManager.ts

```diff
@@ -82,59 +87,54 @@ export class ConnectionManager extends (EventEmitter as {
    * @param tabId - the id of the tab to disconnect
    */
   disconnectTab(tabId: number): void {
-    this.#chainConnections.forEach((a) => {
-      if (a.tabId === tabId) a.port.disconnect()
-    })
+    const tab = this.#tabs.get(tabId)
+    if (!tab) return
+    tab.port.disconnect()
   }
 
   /**
-   * addChainConnection registers a new chain-connection to be tracked by the background.
+   * addTab registers a new tab tracked by the background.
    *
    * @param port - a port for a fresh connection that was made to the background
    * by a content script.
    */
-  addChainConnection(port: chrome.runtime.Port): void {
+  addTab(port: chrome.runtime.Port): void {
     if (!this.#client) {
       throw new Error("Smoldot client does not exist.")
     }
 
     // if create an `AppMediator` throws, it has sent an error down the
     // port and disconnected it, so we should just ignore
     try {
-      const { name: chainId, sender } = port
-      const tabId: number = sender!.tab!.id!
-      const url: string = sender!.url!
-
-      const healthChecker = smHealthChecker()
-      const id = `${chainId}::${tabId}`
-      const connection: ChainConnection = {
-        id,
-        chainName: "",
-        chainId,
-        tabId,
-        url,
-        port,
-        healthChecker,
-      }
+      const tabId: number = port.sender!.tab!.id!
+      const url: string = port.sender!.url!
 
       const onMessageHandler = (msg: ToExtension) => {
-        this.#handleMessage(msg, connection)
+        const tab = this.#tabs.get(tabId)
+        if (tab) this.#handleMessage(msg, tab)
       }
       port.onMessage.addListener(onMessageHandler)
 
       const onDisconnect = () => {
-        connection.chain && connection.chain.remove()
-
         port.onMessage.removeListener(onMessageHandler)
         port.onDisconnect.removeListener(onDisconnect)
 
-        if (this.#chainConnections.get(id)?.chainName) this.#emitStateChanged()
+        const tab = this.#tabs.get(tabId)
+        if (tab) {
+          tab.chainConnections.forEach((chain) => {
+            chain.chain?.remove()
+          })
+        }
 
-        this.#chainConnections.delete(id)
+        this.#emitStateChanged()
+        this.#tabs.delete(tabId)
       }
       port.onDisconnect.addListener(onDisconnect)
 
-      this.#chainConnections.set(id, connection)
+      this.#tabs.set(tabId, {
+        chainConnections: new Map(),
+        port,
+      })
```

Comment:
A way to further simplify this code is to define the `tab` as a clojure, i.e:

```ts
      const tabId: number = port.sender!.tab!.id!
      const url: string = port.sender!.url!
      const tab = { port, chainConnections: new Map() }
      this.#tabs.set(tabId, tab)

      const onMessageHandler = (msg: ToExtension) => {
        this.#handleMessage(msg, tab)
      }
      port.onMessage.addListener(onMessageHandler)

      const onDisconnect = () => {
        port.onMessage.removeListener(onMessageHandler)
        port.onDisconnect.removeListener(onDisconnect)

        tab.chainConnections.forEach((chain) => {
          chain.chain?.remove()
        })
        
        this.#emitStateChanged()
        this.#tabs.delete(tabId)
      }
      port.onDisconnect.addListener(onDisconnect)
```

Reply:
This is not the same thing. I do `this.#tabs.get(tabId)` within callbacks to check whether the tab is still in the list.

Reply:
But the thing is that it will always be in the list, it's impossible that it won't be in the list. That's why I think that the code is a bit misleading.